Library/Homebrew/cmd/pull.rb

```diff
@@ -217,4 +231,141 @@ def pull
+      @patch_url = url + ".patch"
+      @patchpath = HOMEBREW_CACHE + File.basename(patch_url)
+    end
+
+    def pull_patch
+      fetch_patch
+      apply_patch
+    end
+
+    def fetch_patch
+      ohai "Fetching patch"
+      curl patch_url, "-o", patchpath
+    end
+
+    def apply_patch
+      # Applies a patch previously downloaded with fetch_patch()
+      # Deletes the patch file as a side effect, regardless of success
+
+      ohai "Applying patch"
+      patch_args = []
+      # Normally we don't want whitespace errors, but squashing them can break
+      # patches so an option is provided to skip this step.
+      if ARGV.include?("--ignore-whitespace") || ARGV.include?("--clean")
+        patch_args << "--whitespace=nowarn"
+      else
+        patch_args << "--whitespace=fix"
+      end
+
+      # Fall back to three-way merge if patch does not apply cleanly
+      patch_args << "-3"
+      patch_args << patchpath
+
+      begin
+        safe_system "git", "am", *patch_args
+      rescue ErrorDuringExecution
+        if ARGV.include? "--resolve"
+          odie "Patch failed to apply: try to resolve it."
+        else
+          system "git", "am", "--abort"
+          odie "Patch failed to apply: aborted."
+        end
+      ensure
+        patchpath.unlink
+      end
+    end
+  end
+
+  # List files changed by a patch, partitioned in to those that are (probably)
+  # formula definitions, and those which aren't. Only applies to patches on 
+  # Homebrew core or taps, based simply on relative pathnames of affected files.
+  def files_changed_in_patch(patchfile)
+    files = []
+    formulae = []
+    others = []
+    File.foreach(patchfile) do |line|
+      files << $1 if line =~ /^\+\+\+ b\/(.*)/
+    end
+    files.each do |file|
+      if file.start_with?("Library/Formula") || file.match("^[^/]*\.rb")
+        formulae << File.basename(file, ".rb")
+      else
+        others << file
+      end
+    end
+    {:files => files, :formulae => formulae, :others => others}
+  end
+
+  # Get current formula versions without loading formula definition in this process
+  # Returns info as a hash (type => version), for pull.rb's internal use
+  # Uses special key :nonexistent => true for nonexistent formulae
+  def current_versions_from_info_external(formula_name)
+    versions = {}
+    json = Utils.popen_read(HOMEBREW_BREW_FILE, "info", "--json=v1", formula_name)
+    if $?.success?
+      info = Utils::JSON.load(json)
+      [:stable, :devel, :head].each do |vertype|
+        versions[vertype] = info[0]["versions"][vertype.to_s]
+      end
+    else
+      versions[:nonexistent] = true
+    end
+    versions
+  end
+
+  def subject_for_bump(formula, old, new)
+    if old[:nonexistent]
+      # New formula
+      ver_strs = []
+      ver_strs << new[:stable] if new[:stable]
+      ver_strs << "#{new[:devel]} (devel)" if new[:devel]
+      subject = "#{formula.name} #{ver_strs.join(", ")} (new formula)"
```

Comment:
Yeah, looks a bit weird to me too. Using parentheses for sequential clauses at what are effectively different grouping levels rubs me the wrong way a bit. Bit I think it's still readable, and it's for human consumption, and like you say, infrequent. (And still completely computer-parseable once you know " (new formula)" is a special-case suffix.) No better format immediately springs to mind for me (w/o sacrificing regularity etc).